### PR TITLE
[WB-7870] Support for TensorFlow/Keras SavedModel format. 

### PR DESCRIPTION
**Pull request by ayulockin**
Fixes WB-7870

Description
-----------
This PR adds the ability to save custom Keras models, which might include subclassed models or models using custom layers, in the `SavedModel` format. This `SavedModel` is then logged as a W&B artifact.

### Current Behavior

Currently the Keras `WandbCallback` saves the model in the HDF5 format ([#L414](https://github.com/wandb/client/blob/80b47093f6ccbe75be7e4382e0cfd948c868763d/wandb/integration/keras/keras.py#L414)) and tries to log it along with the W&B directory. However, this has limitations:
* HDF5 is not the recommended method to save TensorFlow/Keras models. 
* `SavedModel` format can be directly used by TF Serving, TF Lite, etc. Basically, this feature addition will serve well for industry usage. 
* Currently, when a subclassed model is used with `save_model=True` W&B throws an error message - `Can't save model, h5py returned error:` which might confuse the user. 

### What this PR is doing?

This PR will save the model in the SavedModel format and log it as an artifact. Why artifact?
* The `SavedModel` format creates a directory in the wandb dir. 
* This dir if uploaded as W&B files (which is the current behaviour) cannot be restored using `wandb.restore`. (I tried it and it works with files and not dir)

### How's the behaviour affected?

For a user using `WandbCallback` with `save_model=True`:
* They will see a wandb warning notifying that in a future release we might deprecate the HDF5 model saving method and opt `SavedModel` format. 

image

* The earlier error is now replaced with a warning.

image

* If Sequential or Functional model is used, the best model will be saved both in  `HDF5` as W&B files and `SavedModel` as Artifacts format.

Here's an example [run page](https://wandb.ai/ayush-thakur/keras/runs/3mbf8cy0?workspace=user-ayush-thakur).

Testing
-------
Tested manually. `yea` test. 

Checklist
-------
- https://wandb.atlassian.net/browse/WB-7870
- https://wandb.atlassian.net/browse/WB-7686

### CODE REVIEW

**Comment by ayulockin** [comment]
Hey @dmitryduev, @raubitsj, @kptkin can someone take a look at this PR. Thanks. 
cc: @soumik12345

**Comment by raubitsj** [comment]
Need approval from @tssweeney  or @vwrj

File: wandb/integration/keras/keras.py
Comment (tssweeney) [inline wandb/integration/keras/keras.py:986]: This looks good. In fact, we use the same exact technique in the TF/Keras serializer. Can we put a comment in saying: 
```
TODO: Replace this manual artifact creation with the `log_model` method after `log_model` is released from beta.
```

File: wandb/integration/keras/keras.py
Comment (tssweeney) [inline wandb/integration/keras/keras.py:996]: Probably want to move this to the top of the file?

File: wandb/integration/keras/keras.py
Comment (tssweeney) [inline wandb/integration/keras/keras.py:992]: for the aliases, i don't see the benefit of `best` - since it is always paired with the `latest` in this case. Seems a bit redundant. Also, the `epoch_` may be redundant as well (since the version index is the epoch). If these are both true, i'd like to remove the aliases here.

**Review by tssweeney — APPROVED** [review]
Approving - assuming the comments are addressed. This looks good and can be easily upgraded to work with the new model reg APIs when we are ready.

File: functional_tests/keras/keras_subclassed_test_tf_2_6.yea
Comment (dmitryduev) [inline functional_tests/keras/keras_subclassed_test_tf_2_6.yea:18]:
```suggestion
    - :wandb:runs[0][exitcode]: 0

```

File: functional_tests/keras/test_keras_subclassed_model.py
Comment (dmitryduev) [inline functional_tests/keras/test_keras_subclassed_model.py:8]: How about using a fixed seed here and then explicitly checking the contents of the saved model/artifact in a few asserts at the end of this script? Just thinking, this might be an overkill.

File: functional_tests/keras/keras_subclassed_test_tf_2_6.yea
Comment (dmitryduev) [reply functional_tests/keras/keras_subclassed_test_tf_2_6.yea:18]: You probably know this, but this flag is very useful when coming up with yea tests, so just throwing it here:
`yea --debug run functional_tests/keras/keras_subclassed_test_tf_2_6.yea`

File: wandb/integration/keras/keras.py
Comment (dmitryduev) [inline wandb/integration/keras/keras.py:412]: If you're think we should plan on deprecating this, please deprecate this following this guide: https://github.com/wandb/client/blob/master/CONTRIBUTING.md#deprecating-features

File: wandb/integration/keras/keras.py
Comment (dmitryduev) [inline wandb/integration/keras/keras.py:1001]: Would you mind adding an explicit unit test to beef up codecov here?

**Review by dmitryduev — COMMENTED** [review]
Looks great! A couple minor comments.
As a side note, I really wish this was available a couple years back when I was using SavedModel's - would have saved me prob months of trouble!

File: wandb/integration/keras/keras.py
Comment (ayulockin) [reply wandb/integration/keras/keras.py:992]: You are right, I too don't see a benefit of `best` as an alias. However, I think `epoch_` is still useful. 

The method `_save_model_as_artifact` is called when `if self.current and self.monitor_op(self.current, self.best):` is true. While training a model this condition might not be true every time and the model will be saved as an artifact at those (where it's true) epochs. Thoughts?

File: functional_tests/keras/keras_subclassed_test_tf_2_6.yea
Comment (ayulockin) [reply functional_tests/keras/keras_subclassed_test_tf_2_6.yea:18]: Thanks for reminding me about this flag.

File: functional_tests/keras/test_keras_subclassed_model.py
Comment (ayulockin) [reply functional_tests/keras/test_keras_subclassed_model.py:8]: When I use the `--debug` flag, I get this dict:
```
INFO: exit= 0
{ ':wandb:artifacts': {'model-lovely-dawn-32': {'num': 2, 'type': 'model'}},
  ':wandb:portfolio_links': {},
  ':wandb:runs': [ { 'alerts': [],
                     'config': {},
                     'config_wandb': { 'cli_version': '0.12.16.dev1',
                                       'code_path': 'code/functional_tests/keras/test_keras_subclassed_model.py',
                                       'framework': 'keras',
                                       'is_jupyter_run': False,
                                       'is_kaggle_kernel': False,
                                       'python_version': '3.7.12',
                                       'start_time': 1650982985,
                                       't': { '1': [2, 3, 55],
                                              '2': [2, 3, 55],
                                              '3': [3, 8],
                                              '4': '3.7.12',
                                              '5': '0.12.16.dev1',
                                              '8': [5]}},
                     'exitcode': 0,
                     'files': { 'code/functional_tests/keras/test_keras_subclassed_model.py': { 'size': 942},
                                'conda-environment.yaml': {'size': 12070},
                                'config.yaml': {'size': 454},
                                'diff.patch': {'size': 1362},
                                'media/graph/graph_summary_c9172717e0c0fa694ed8.graph.json': { 'size': 381},
                                'output.log': {'size': 2466},
                                'requirements.txt': {'size': 4728},
                                'wandb-metadata.json': {'size': 753},
                                'wandb-summary.json': {'size': 470}},
                     'git': { 'commit': 'c8f3625ed8c0d468d41d51c217e5f4458c552f6f',
                              'remote': 'https://github.com/soumik12345/client'},
                     'group': '',
                     'history': [ { '_runtime': 18,
                                    '_step': 0,
                                    '_timestamp': 1650983003,
                                    'accuracy': 0.10000000149011612,
                                    'epoch': 0,
                                    'loss': 247.49560546875,
                                    'val_accuracy': 0.15000000596046448,
                                    'val_loss': 2.302488327026367},
                                  { '_runtime': 19,
                                    '_step': 1,
                                    '_timestamp': 1650983004,
                                    'accuracy': 0.10999999940395355,
                                    'epoch': 1,
                                    'loss': 2.302546501159668,
                                    'val_accuracy': 0.15000000596046448,
                                    'val_loss': 2.3022634983062744}],
                     'host': 'ayut-simple',
                     'job_type': '',
                     'metrics': {},
                     'name': '',
                     'notes': '',
                     'output': { 'stderr': [ '2022-04-26 14:23:12.146581: W '
                                             'tensorflow/stream_executor/platform/default/dso_loader.cc:64] '
                                             'Could not load dynamic library '
                                             "'libcuda.so.1'; dlerror: "
                                             'libcuda.so.1: cannot open shared '
                                             'object file: No such file or '
                                             'directory; LD_LIBRARY_PATH: '
                                             '/usr/local/cuda/lib64:/usr/local/nccl2/lib:/usr/local/cuda/extras/CUPTI/lib64',
                                             '2022-04-26 14:23:12.146646: W '
                                             'tensorflow/stream_executor/cuda/cuda_driver.cc:269] '
                                             'failed call to cuInit: UNKNOWN '
                                             'ERROR (303)',
                                             '2022-04-26 14:23:12.146672: I '
                                             'tensorflow/stream_executor/cuda/cuda_diagnostics.cc:156] '
                                             'kernel driver does not appear to '
                                             'be running on this host '
                                             '(ayut-simple): '
                                             '/proc/driver/nvidia/version does '
                                             'not exist',
                                             '2022-04-26 14:23:12.147201: I '
                                             'tensorflow/core/platform/cpu_feature_guard.cc:142] '
                                             'This TensorFlow binary is '
                                             'optimized with oneAPI Deep '
                                             'Neural Network Library (oneDNN) '
                                             'to use the following CPU '
                                             'instructions in '
                                             'performance-critical '
                                             'operations:  AVX2 FMA',
                                             'To enable them in other '
                                             'operations, rebuild TensorFlow '
                                             'with the appropriate compiler '
                                             'flags.',
                                             '\x1b[34m\x1b[1mwandb\x1b[39m\x1b[22m: '
                                             '\x1b[33mWARNING\x1b[39m The '
                                             'save_model argument by default '
                                             'saves the model in the HDF5 '
                                             'format that cannot save custom '
                                             'objects like subclassed models '
                                             'and custom layers. This behavior '
                                             'will be deprecated in a future '
                                             'release in favor of the '
                                             'SavedModel format. Meanwhile, '
                                             'the HDF5 model is saved as W&B '
                                             'files and the SavedModel as W&B '
                                             'Artifacts.',
                                             '2022-04-26 14:23:21.161710: I '
                                             'tensorflow/compiler/mlir/mlir_graph_optimization_pass.cc:185] '
                                             'None of the MLIR Optimization '
                                             'Passes are enabled (registered '
                                             '2)',
                                             '\x1b[34m\x1b[1mwandb\x1b[39m\x1b[22m: '
                                             '\x1b[32m\x1b[41mERROR\x1b[39m\x1b[49m '
                                             "Can't save model in the h5py "
                                             'format. The model will be saved '
                                             'as W&B Artifacts in the '
                                             'SavedModel format.',
                                             '2022-04-26 14:23:24.067492: W '
                                             'tensorflow/python/util/util.cc:348] '
                                             'Sets are not currently '
                                             'considered sequences, but this '
                                             'may change in the future, so '
                                             'consider avoiding using them.',
                                             '\x1b[34m\x1b[1mwandb\x1b[39m\x1b[22m: '
                                             'Adding directory to artifact '
                                             '(/home/ayusht/wandb-cli/client/functional_tests/keras/wandb/run-20220426_142305-2catnaoo/files/model-best)... '
                                             'Done. 0.1s',
                                             '\x1b[34m\x1b[1mwandb\x1b[39m\x1b[22m: '
                                             'Adding directory to artifact '
                                             '(/home/ayusht/wandb-cli/client/functional_tests/keras/wandb/run-20220426_142305-2catnaoo/files/model-best)... '
                                             'Done. 0.1s'],
                                 'stdout': [ 'Epoch 1/2',
                                             '4/4 '
                                             '[==============================] '
                                             '- 2s 71ms/step - loss: 247.4956 '
                                             '- accuracy: 0.1000 - val_loss: '
                                             '2.3025 - val_accuracy: 0.1500',
                                             'Epoch 2/2',
                                             '4/4 '
                                             '[==============================] '
                                             '- 0s 16ms/step - loss: 2.3025 - '
                                             'accuracy: 0.1100 - val_loss: '
                                             '2.3023 - val_accuracy: 0.1500']},
                     'program': './test_keras_subclassed_model.py',
                     'run_id': '2catnaoo',
                     'summary': { 'accuracy': 0.10999999940395355,
                                  'best_epoch': 1,
                                  'best_val_loss': 2.3022634983062744,
                                  'epoch': 1,
                                  'graph': { '_type': 'graph-file',
                                             'path': 'media/graph/graph_summary_c9172717e0c0fa694ed8.graph.json',
                                             'sha256': 'c9172717e0c0fa694ed800293d3019e42127def001df804e051ce98464ef8707',
                                             'size': 381},
                                  'loss': 2.302546501159668,
                                  'val_accuracy': 0.15000000596046448,
                                  'val_loss': 2.3022634983062744},
                     'summary_wandb': {'runtime': 20},
                     'tags': [],
                     'telemetry': { '1': [2, 3, 55],
                                    '2': [2, 3, 55],
                                    '3': [3, 8],
                                    '4': '3.7.12',
                                    '5': '0.12.16.dev1',
                                    '8': [5]}}],
  ':wandb:runs_len': 1,
  ':wandb:sentry_events': [],
  ':yea:exit': 0}
```

Not sure if I can check the content of the saved model/artifact. The only info that I have access to imo is: { ':wandb:artifacts': {'model-lovely-dawn-32': {'num': 2, 'type': 'model'}},`.

Comment (dmitryduev) [reply functional_tests/keras/test_keras_subclassed_model.py:8]: I was thinking about explicit assert statements in the training script itself, as in read in the resulting model and check what's in there. But like i said, this might be an overkill -- up to you to decide whether to add that (if you see value in it).

File: wandb/integration/keras/keras.py
Comment (ayulockin) [reply wandb/integration/keras/keras.py]: I don't think this except block is useful here. Removing it actually. 

If you defer can you let me know what possible exception it can hit? `ImportError` is something it can never hit imo. Not very sure about `RuntimeError`.

File: functional_tests/keras/test_keras_subclassed_model.py
Comment (ayulockin) [reply functional_tests/keras/test_keras_subclassed_model.py:10]: Thanks for the clarification here. I have modified the test to assert on the files logged as artifact. Can you please take a look? :)

**Comment by ayulockin** [comment]
Hey @dmitryduev whenever you get a chance do let me know what else is required from my end. :)

**Review by dmitryduev — APPROVED** [review]

**Comment by dmitryduev** [comment]
@ayulockin seeing the relevant test flaking out, please look into it: https://app.circleci.com/pipelines/github/wandb/client/12213/workflows/55281dda-88c5-45aa-a767-fb3565455ad4/jobs/177362

**Review by dmitryduev — APPROVED** [review]
Approving and merging with a note to our future selves:

The new test added in this PR flaked out on one occasion, in a way that we've observed on a number of occasions in artifact-related tests: https://app.circleci.com/pipelines/github/wandb/client/12213/workflows/55281dda-88c5-45aa-a767-fb3565455ad4/jobs/177362
We should investigate the causes and fix the flake, but it is outside the scope of this PR.

**Comment by dmitryduev** [comment]
Thanks @ayulockin !!

**Comment by ayulockin** [comment]
I was about to work on the flaky and saw your note for future selves. I wasn't very sure why it had failed. 

Thanks for merging it @dmitryduev. :)

**Pull request merged**
